packages/common/src/services/audius-backend/solana.ts

```diff
@@ -0,0 +1,254 @@
+  }
+) => {
+  return (
+    await audiusBackendInstance.getAudiusLibs()
+  ).solanaWeb3Manager!.getTokenAccountInfo(tokenAccount.toString(), mint)
+}
+
+export const deriveUserBankPubkey = async (
+  audiusBackendInstance: AudiusBackend,
+  { ethAddress, mint = DEFAULT_MINT }: UserBankConfig = {}
+) => {
+  const audiusLibs: AudiusLibs = await audiusBackendInstance.getAudiusLibs()
+  return await audiusLibs.solanaWeb3Manager!.deriveUserBank({
+    ethAddress,
+    mint
+  })
+}
+
+export const deriveUserBankAddress = async (
+  audiusBackendInstance: AudiusBackend,
+  { ethAddress, mint = DEFAULT_MINT }: UserBankConfig = {}
+) => {
+  const pubkey = await deriveUserBankPubkey(audiusBackendInstance, {
+    ethAddress,
+    mint
+  })
+  return pubkey.toString() as SolanaWalletAddress
+}
+
+type CreateUserBankIfNeededConfig = UserBankConfig & {
+  recordAnalytics: (event: AnalyticsEvent, callback?: () => void) => void
+  feePayerOverride: string
+}
+
+type CreateUserBankIfNeededErrorResult = {
+  error: string
+  errorCode: string | number | null
+}
+type CreateUserBankIfNeededSuccessResult = {
+  didExist: boolean
+  userbank: PublicKey
+}
+type CreateUserBankIfNeededResult =
+  | CreateUserBankIfNeededSuccessResult
+  | CreateUserBankIfNeededErrorResult
+
+function isCreateUserBankIfNeededError(
+  res: CreateUserBankIfNeededResult
+): res is CreateUserBankIfNeededErrorResult {
+  return 'error' in res
+}
+
+/**
+ * Attempts to create a userbank if one does not exist.
+ * Defaults to AUDIO mint and the current user's wallet.
+ */
+export const createUserBankIfNeeded = async (
+  audiusBackendInstance: AudiusBackend,
+  {
+    recordAnalytics,
+    feePayerOverride,
+    mint = DEFAULT_MINT,
+    ethAddress
+  }: CreateUserBankIfNeededConfig
+) => {
+  const audiusLibs: AudiusLibs = await audiusBackendInstance.getAudiusLibs()
+
+  const recipientEthAddress =
+    ethAddress ?? audiusLibs.Account!.getCurrentUser()?.wallet
+
+  if (!recipientEthAddress) {
+    throw new Error(
+      `createUserBankIfNeeded: Unexpectedly couldn't get recipient eth address`
+    )
+  }
+
+  try {
+    const res: CreateUserBankIfNeededResult =
+      await audiusLibs.solanaWeb3Manager!.createUserBankIfNeeded({
+        feePayerOverride,
+        ethAddress: recipientEthAddress,
+        mint
+      })
+
+    if (isCreateUserBankIfNeededError(res)) {
+      // Will catch and log below
+      throw res
+    }
+
+    // If it already existed, return early
+    if (res.didExist) {
+      console.log('Userbank already exists')
+    } else {
+      // Otherwise we must have tried to create one
+      console.info(`Userbank doesn't exist, attempted to create...`)
+
+      recordAnalytics({
+        eventName: Name.CREATE_USER_BANK_REQUEST,
+        properties: { mint, recipientEthAddress }
+      })
+    }
+
+    recordAnalytics({
+      eventName: Name.CREATE_USER_BANK_SUCCESS,
+      properties: { mint, recipientEthAddress }
+    })
+    return res.userbank
```

Comment:
@rickyrombo I updated this to correctly return the user bank from the inner function!
Also reorganized a bit to make it less complicated and it will actually throw its error now

Reply:
wanna make sure tipping works on stage after this!

Reply:
I tested it locally and it worked fine. But we definitely should re-verify after merging and during RC